plugins/genesys/fu-genesys-usbhub-device.c

```diff
@@ -1134,32 +1380,296 @@ static void
 fu_genesys_usbhub_device_to_string(FuDevice *device, guint idt, GString *str)
 {
 	FuGenesysUsbhubDevice *self = FU_GENESYS_USBHUB_DEVICE(device);
-	fu_string_append_kx(str, idt, "FlashEraseDelay", self->flash_erase_delay);
-	fu_string_append_kx(str, idt, "FlashWriteDelay", self->flash_write_delay);
+	guint64 fw_max_size = fu_device_get_firmware_size_max(device);
+	guint idt_bank = idt + 1;
+	guint idt_bank_detail = idt_bank + 1;
+
+	fu_string_append_ku(str, idt, "FlashEraseDelay", self->flash_erase_delay);
+	fu_string_append_ku(str, idt, "FlashWriteDelay", self->flash_write_delay);
 	fu_string_append_kx(str, idt, "FlashBlockSize", self->flash_block_size);
 	fu_string_append_kx(str, idt, "FlashSectorSize", self->flash_sector_size);
 	fu_string_append_kx(str, idt, "FlashRwSize", self->flash_rw_size);
-	fu_string_append_kx(str,
-			    idt,
-			    "FwBank0Addr",
-			    self->spec.fw_bank_addr[FW_BANK_1][FW_TYPE_HUB]);
-	fu_string_append_kx(str, idt, "FwBank0Vers", self->fw_bank_vers[FW_BANK_1][FW_TYPE_HUB]);
-	if (fu_device_has_flag(device, FWUPD_DEVICE_FLAG_DUAL_IMAGE)) {
-		fu_string_append_kx(str,
-				    idt,
-				    "FwBank1Addr",
-				    self->spec.fw_bank_addr[FW_BANK_2][FW_TYPE_HUB]);
+
+	fu_string_append(str,
+			 idt,
+			 "RunningBank",
+			 fu_genesys_fw_status_to_string(self->running_bank));
+	fu_string_append_kb(str, idt, "SupportDualBank", self->spec.support_dual_bank);
+	fu_string_append_kb(str, idt, "SupportCodeSize", self->spec.support_code_size);
+
+	for (gint i = 0; i < FU_GENESYS_FW_TYPE_INSIDE_HUB_COUNT; i++) {
+		if (i == FU_GENESYS_FW_TYPE_CODESIGN ||	   /* do not show codesign info */
+		    self->spec.fw_bank_capacity[i] == 0 || /* unsupported fw type */
+		    fw_max_size <= self->spec.fw_bank_addr[FW_BANK_1][i]) /* capacity too small */
+			continue;
```

Comment:
maybe split out the contents to something like `fu_genesys_usbhub_device_codesign_to_string()`? Then you can unindent everything a bit.